src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -989,82 +946,153 @@ HeapWord* G1CollectedHeap::satisfy_failed_allocation_helper(size_t word_size,
          "Flag should have been handled and cleared prior to this point");
+  return false;
+}
 
-  // What else?  We might try synchronous finalization later.  If the total
-  // space available is large enough for the allocation, then a more
-  // complete compaction phase than we've tried so far might be
-  // appropriate.
-  return nullptr;
+void G1CollectedHeap::reset_allocation_requests() {
+  assert_at_safepoint_on_vm_thread();
+
+  DoublyLinkedList<StalledAllocReq> prev_allocations;
+
+  prev_allocations.swap(_satisfied_allocations);
+
+  assert(_satisfied_allocations.is_empty(), "Sanity check!");
+
+  DoublyLinkedList<StalledAllocReq>::RemoveIterator iter(&prev_allocations);
+
+  for (StalledAllocReq* alloc_req; iter.next(&alloc_req);) {
+    if (alloc_req->state() == StalledAllocReq::AllocationState::Failed) {
+      // If an allocation request was declared failed, we maintain the failed state.
+      // This prevents requests from flip-floping between Pending->Failed->Pending states.
+      // Once any allocation attempt declares a request as failed, that request will not be re-attempted.
+      _satisfied_allocations.insert_first(alloc_req);
+    } else {
+      alloc_req->set_state(StalledAllocReq::AllocationState::Pending);
+      _stalled_allocations.insert_first(alloc_req);
+    }
+  }
 }
 
-// Attempting to expand the heap sufficiently
-// to support an allocation of the given "word_size".  If
-// successful, perform the allocation and return the address of the
-// allocated block, or else null.
+bool G1CollectedHeap::handle_allocation_requests(bool expect_null_mutator_alloc_region) {
+  assert_at_safepoint_on_vm_thread();
 
-HeapWord* G1CollectedHeap::expand_and_allocate(size_t word_size) {
+  const uint active_numa_nodes = G1NUMA::numa()->num_active_nodes();
+  bool *expect_null_alloc_regions = (bool *)alloca(active_numa_nodes * sizeof(bool));
+  for (uint i = 0; i < active_numa_nodes; i++) {
+    expect_null_alloc_regions[i] = expect_null_mutator_alloc_region;
+  }
+
+  while(true) {
+    StalledAllocReq* alloc_req = _stalled_allocations.first();
+    if (alloc_req == nullptr) {
+      // No more pending requests, all allocations succeeded
+      return true;
+    }
+
+    HeapWord* result =
+      satisfy_failed_allocation_helper(alloc_req->size(),
+                                       alloc_req->node_index(),
+                                       expect_null_alloc_regions[alloc_req->node_index()]
+                                       );
+
+    if (result == nullptr) {
+      // Failed to allocate, give up.
+      return false;
+    }
+
+    expect_null_alloc_regions[alloc_req->node_index()] = false;
+
+    // Allocation succeeded, update the state and result of the allocation request
+    alloc_req->set_state(StalledAllocReq::AllocationState::Success, result);
+
+    if (is_humongous(alloc_req->size())) {
+      // Calculate payload size and initialize the humongous object with a fillerArray
```

Comment:
```suggestion
      // Calculate payload size and initialize the humongous object with a fillerArray.
```

It is nowhere explained why this is needed.